csrc/id_model/id_model.h

```diff
@@ -0,0 +1,136 @@
+//
+// Exact mode is if the iter domain relationships from producer to consumer are
+// considered the exact same size operating on matching dimensions from the root
+// domain mapping.
+//
+// IdMappingMode::EXACT
+//   Don't map any broadcast axes to non-broadcast axes
+//   Do not forward through any broadcast IDs
+//
+class IdModel : public PolymorphicBase {
+ public:
+  IdModel(
+      const std::vector<Expr*>& exprs,
+      const std::vector<TensorView*>& additional_tvs);
+
+  IdModel(const std::vector<Expr*>& exprs);
+
+  // Same as the above constructor with fusion->exprs() excpet fusion may have
+  // some dangling inputs/outputs that are expected to have IterDomain entries
+  // even though there's no possible connections from them.
+  IdModel(Fusion* fusion);
+
+  // Returns iter domain graph of provided mode.
+  const ValGraph& idGraph(IdMappingMode mode) const;
+  ValGraph& idGraph(IdMappingMode mode);
+
+  // IterDomains from the original fusion are only allowed to be used once in
+  // the IterDomain graph, id->uses() are not directly used as there's no bounds
+  // check that would prevent a use from being defined that's not part of the
+  // actual fusion definition.
+  //
+  // Note, any iter domains used during something like loop or concrete id
+  // resolution could actually have multiple Expr* uses, and uses on disjoint id
+  // sets should be used, not this.
+  //
+  // TODO: Refactor or remove?
+  Expr* idUse(IterDomain* id) const;
+  Expr* idDef(IterDomain* id) const;
+
+  // TODO: Seems a bit unfortunate that this isn't IterDomain local information.
+  const std::unordered_set<IterDomain*>& viewRfactorIds() const {
+    return view_rfactor_ids_;
+  }
+
+  // Update the LOOP ID disjoint sets with resolved computeWith
+  void updateComputeWith(TensorView* compute_with_tv);
+
+  std::string toString() const;
+
+  // TODO: Should this not be private?
+ protected:
+  // Sometimes fusion inputs or outputs are disconnected from expressions, in
+  // those cases we still may want to send in some additional tensor views from
+  // the Fusion that don't have expressions associated with them.
+  void build(
+      const std::vector<Expr*>& exprs,
+      const std::vector<TensorView*>& additional_tvs);
+
+  // ======= START Iteration domain build process in order called =======
+
+  // Fills id_uses_ and id_definitions_ for all IterDomains active in the
+  // fusion.
+  void buildIterDomainDefinitionsAndUses(
+      const std::vector<TensorView*>& all_tvs);
+
+  // Iterates over all IterDomains in id_definitions_ and calls initializeID on
+  // a new IdGraph and returns it.
+  ValGraph initializeIdGraph(bool propagate_through_exprs = true);
+
+  // Fills disjoint_ids_[IdMappingMode::EXACT] for relationships between inputs
+  // and first output of expr
+  void buildExactGraph(const std::vector<Expr*>& exprs);
+
+  // Keeps a disjoint set entry for all IterDomain for all mapping mode types.
```

Comment:
Is this comment for `id_graphs_`? Should it be
```suggestion
  // Keeps a ValGraph containing all IterDomain for all mapping mode types.
```